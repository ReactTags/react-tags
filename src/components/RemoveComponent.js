import React from 'react';
import PropTypes from 'prop-types';

const crossStr = String.fromCharCode(215);
const RemoveComponent = (props) => {
  const { readOnly, removeComponent, onClick, className, tag, index } = props;
  if (readOnly) {
    return <span />;
  }

  if (removeComponent) {
    const Component = removeComponent;
    return <Component {...props} />;
  }

  return (
    <button onClick={onClick} className={className} aria-label={`Tag at index ${index} with value ${tag.id} focussed. Press enter to remove`}>
      {crossStr}
    </button>
  );
};

RemoveComponent.propTypes = {
  className: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  readOnly: PropTypes.bool,
  removeComponent: PropTypes.func,
  tag: PropTypes.shape({
    id: PropTypes.string.isRequired,
    className: PropTypes.string,
    key: PropTypes.string,
  }),
  index: PropTypes.number.isRequired,
};

export default RemoveComponent;
