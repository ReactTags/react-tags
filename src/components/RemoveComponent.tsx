import React from 'react';
import { KEYS } from './constants';
import { Tag } from './SingleTag';

const crossStr = String.fromCharCode(215);

export interface RemoveComponentProps {
  onRemove: (
    event:
      | React.MouseEvent<HTMLButtonElement>
      | React.KeyboardEvent<HTMLButtonElement>
  ) => void;
  readOnly: boolean;
  removeComponent?: React.ComponentType<any>;
  className?: string;
  tag: Tag;
  index: number;
}

const RemoveComponent = (props: RemoveComponentProps) => {
  const { readOnly, removeComponent, onRemove, className, tag, index } = props;

  const onKeydown = (event: React.KeyboardEvent<HTMLButtonElement>) => {
    if (KEYS.ENTER.includes(event.keyCode) || event.keyCode === KEYS.SPACE) {
      event.preventDefault();
      event.stopPropagation();
      return;
    }
    if (event.keyCode === KEYS.BACKSPACE) {
      onRemove(event);
    }
  };

  if (readOnly) {
    return <span />;
  }

  const ariaLabel = `Tag at index ${index} with value ${tag.id} focussed. Press backspace to remove`;
  if (removeComponent) {
    const Component = removeComponent;
    return (
      <Component
        data-testid="remove"
        onRemove={onRemove}
        onKeyDown={onKeydown}
        className={className}
        aria-label={ariaLabel}
        tag={tag}
        index={index}
      />
    );
  }

  return (
    <button
      data-testid="remove"
      onClick={onRemove}
      onKeyDown={onKeydown}
      className={className}
      type="button"
      aria-label={ariaLabel}>
      {crossStr}
    </button>
  );
};

export default RemoveComponent;
