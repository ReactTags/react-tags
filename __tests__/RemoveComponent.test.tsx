import React from 'react';
import { stub } from 'sinon';

import RemoveComponent, {
  RemoveComponentProps,
} from '../src/components/RemoveComponent';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Test <RemoveComponent/>', () => {
  let minProps, onRemoveStub;
  beforeAll(() => {
    onRemoveStub = stub();
    minProps = {
      onRemove: onRemoveStub,
      index: 0,
      tag: {
        id: 'Tags',
        key: 'Tags',
      },
    };
  });

  function mockItem(override?: RemoveComponentProps) {
    const props = Object.assign({}, minProps, override);
    return <RemoveComponent {...props} />;
  }

  it('should render with base structure', () => {
    const { container } = render(<RemoveComponent {...minProps} />);
    jestExpect(container).toMatchSnapshot();
  });

  it('should return empty span tag when readOnly set to true', () => {
    const props = {
      onRemove: stub(),
      index: 0,
      readOnly: true,
      tag: {
        id: 'Tags',
        key: 'Tags',
        className: '',
      },
    };
    const { container } = render(<RemoveComponent {...props} />);
    jestExpect(container).toMatchInlineSnapshot(`
      <div>
        <span />
      </div>
    `);
  });

  it('should run onRemove when backspace is pressed', () => {
    render(mockItem());
    const removeComp = screen.getByTestId('remove');
    fireEvent.click(removeComp);
    expect(onRemoveStub.calledOnce).toBeTruthy();
  });

  it('should render passed in remove component', () => {
    const CustomRemoveComponent = () => <a className="remove">delete me</a>;
    const { container } = render(
      mockItem({ removeComponent: CustomRemoveComponent })
    );
    jestExpect(container).toMatchInlineSnapshot(`
      <div>
        <a
          class="remove"
        >
          delete me
        </a>
      </div>
    `);
  });
});
