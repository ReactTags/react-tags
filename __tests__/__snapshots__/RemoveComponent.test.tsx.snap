// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <RemoveComponent/> should render with base structure 1`] = `
<div>
  <button
    aria-label="Tag at index 0 with value Tags focussed. Press backspace to remove"
    data-testid="remove"
    type="button"
  >
    Ã—
  </button>
</div>
`;
